Strict mode for Date validator

michalbundyra opened the pull request:
Input must be always in the defined format and must be the same as output
of format method of DateTime.

Strict mode is set to `false` by default to keep BC.

Currently to the validator we can provide `integer`, `double`, `array`, `string` or `DateTimeInterface` value. In my opinion as validator has set format we should check if the input value has exactly the same format as set.
In strict mode we are comparing if:
```php
$input === DateTime::createFromFormat($format, $input)->format($format);
```

Resolve #33
Fix zendframework/zendframework#6407

- [x] Are you creating a new feature?
  - [x] Why is the new feature needed? What purpose does it serve?
  - [x] How will users use the new feature?
  - [x] Base your feature on the `develop` branch, and submit against that branch.
  - [x] Add only one feature per pull request; split multiple features over multiple pull requests
  - [x] Add tests for the new feature.
  - [ ] Add documentation for the new feature.
  - [ ] Add a `CHANGELOG.md` entry for the new feature.

I would see strict mode by default enabled in v3, or ONLY strict mode in v3.
Thoughts?

/cc @svycka @gianarb

michalbundyra committed:
Strict mode for Date validator

Input must be always in the defined format and must be the same as output
of format method of DateTime.

Strict mode is set to `false` by default to keep BC.

Resolve #33
Fix zendframework/zendframework#6407

svycka reviewed (commented):
maybe in v3, we could leave the only strict mode?

webimpress commented on src/Date.php:
```diff
@@ -100,6 +105,32 @@ public function setFormat($format = self::FORMAT_DEFAULT)
+     * @param bool $strict
+     * @return $this
+     */
+    public function setStrict($strict)
```
Not in v2.

michalbundyra committed:
Adds example with Unix timestamp format

- per @svycka

webimpress commented on test/DateTest.php:
```diff
@@ -44,48 +44,53 @@ public function testSetFormatIgnoresNull()
+            ['6',                       'd',               true,     false],
+            ['06',                      'd',               true,     true],
+            [123,                       null,              true,     false],
+            [1340677235,                null,              true,     false],
```
The problem here is that input argument must be string even for `U` format. Output of `DateTime::format` is __always__ a string.

froschdesign reviewed (changes requested):
The documentation for the _new_ strict mode is needed. The following points should be included:

* Why is the strict mode required?
* Which mode should be preferred?
* And like always: code examples

weierophinney committed:
qa: use 7.1 features on new code

Updates new methods of Date validator and its tests to use typehints.

Updates license docblock year range.

weierophinney committed:
docs: documents Date validator strict mode

weierophinney reviewed (approved):
Since #264, the develop branch now targets PHP 7.1. As such, I've pushed changes to add typehints to the newly introduced methods, which will prevent BC breaks in the future, and simplifies their implementation.

I've also added documentation on the new strict mode.
